packages/angular_devkit/architect/src/architect.ts

```diff
@@ -0,0 +1,262 @@
+  constructor(builder: string) {
+    super(`Builder '${builder}' cannot be resolved.`);
+  }
+}
+
+export class WorkspaceNotYetLoadedException extends BaseException {
+  constructor() { super(`Workspace needs to be loaded before Architect is used.`); }
+}
+
+export interface Target<OptionsT = {}> {
+  root: Path;
+  projectType: string;
+  builder: string;
+  options: OptionsT;
+}
+
+export class Architect {
+  private _workspaceSchema = pathResolve(__dirname, 'workspace-schema.json');
+  private _buildersSchema = pathResolve(__dirname, 'builders-schema.json');
+  private _workspace: Workspace;
+
+  constructor(private _root: Path, private _host: virtualFs.Host<{}>) { }
+
+  loadWorkspaceFromHost(workspacePath: Path) {
+    return this._host.read(join(this._root, workspacePath)).pipe(
+      concatMap((buffer) => {
+        const json = JSON.parse(virtualFs.fileBufferToString(buffer));
+
+        return this.loadWorkspaceFromJson(json);
+      }),
+    );
+  }
+
+  loadWorkspaceFromJson(json: Workspace) {
+    return this._validateAgainstSchema(json, this._workspaceSchema).pipe(
+      concatMap((validatedWorkspace: Workspace) => {
+        this._workspace = validatedWorkspace;
+
+        return of(this);
+      }),
+    );
+  }
+
+  getTarget<OptionsT>(
+    projectName?: string,
+    targetName?: string,
+    configurationName?: string,
+    additionalInput?: Partial<OptionsT>,
+  ): Target<OptionsT> {
+    if (!this._workspace) {
+      throw new WorkspaceNotYetLoadedException();
+    }
+
+    projectName = projectName || this._workspace.defaultProject as string;
+    const project = this._workspace.projects[projectName];
+
+    if (!project) {
+      throw new ProjectNotFoundException(projectName);
+    }
+
+    targetName = targetName || project.defaultTarget as string;
+    const projectTarget = project.targets[targetName];
+
+    if (!projectTarget) {
+      throw new TargetNotFoundException(targetName);
+    }
+
+    const options = projectTarget.options;
+    let configuration;
+
+    if (configurationName) {
+      configuration = projectTarget.configurations
+        && projectTarget.configurations[configurationName];
+
+      if (!configuration) {
+        throw new ConfigurationNotFoundException(configurationName);
+      }
+    }
+
+    // Resolve root for the target.
+    // TODO: add Path format to JSON schemas
+    const target: Target<OptionsT> = {
+      root: resolve(this._root, normalize(project.root)),
+      projectType: project.projectType,
+      builder: projectTarget.builder,
+      options: {
+        ...options,
+        ...configuration ? configuration : {},
+        ...additionalInput ? additionalInput as {} : {},
+      } as OptionsT,
+    };
+
+    return target;
+  }
+
+  // Will run the target using the target.
+  run<OptionsT>(
+    target: Target<OptionsT>, logger?: logging.Logger, stop?: Observable<void>,
+  ): Observable<BuildEvent> {
+    logger = logger || createConsoleLogger();
```

Comment:
Should the default be a null logger instead of console?  Not specifying a logger, to me at least, would indicate I do not want logging.

Reply:
Hm... that's debatable IMHO. I think it would generally be surprising to not have any output with the defaults. @hansl WDYT?

Reply:
Yes. `NullLogger` are preferred IMO. Same with Schematics.

Reply:
Done.